File: solver.py
VARIABLES = ['A', 'B', 'C', 'D']
N_VARS = 4
CELL_COUNT = 1 << N_VARS


def to_bin(n):
    return format(n, f'0{N_VARS}b')


def get_term_name(pattern):
    parts = []
    for i, ch in enumerate(pattern):
        if ch == '1':
            parts.append(VARIABLES[i])
        elif ch == '0':
            parts.append(VARIABLES[i] + "'")
    return ''.join(parts) if parts else '1'


def combine_terms(t1, t2):
    diff_idx = None
    result = []
    for i in range(N_VARS):
        a, b = t1[i], t2[i]
        if a == '-' and b == '-':
            result.append('-')
        elif a == '-' or b == '-':
            return None
        elif a == b:
            result.append(a)
        else:
            if diff_idx is not None:
                return None
            diff_idx = i
            result.append('-')
    if diff_idx is not None:
        return ''.join(result)
    return None


def build_kmap(minterms, dont_cares):
    rows = ['00', '01', '11', '10']
    kmap = []
    for r in rows:
        for c in rows:
            bits = r + c
            idx = int(bits, 2)
            if idx in dont_cares:
                val = 'x'
            else:
                val = 1 if idx in minterms else 0
            kmap.append({'index': idx, 'binary': bits, 'value': val, 'row': r, 'col': c})
    return kmap


def simplify(minterms, dont_cares=None):
    if dont_cares is None:
        dont_cares = []

    all_terms = sorted(set(minterms + dont_cares))
    actual = set(minterms)

    if not actual:
        return {
            'expression': '0',
            'kmap': build_kmap(minterms, dont_cares),
            'steps': [{'title': 'No Minterms', 'text': 'No minterms selected — output is 0'}],
            'groups': []
        }

    if len(actual) == CELL_COUNT:
        return {
            'expression': '1',
            'kmap': build_kmap(minterms, dont_cares),
            'steps': [{'title': 'All Minterms', 'text': 'All cells are 1 — output is 1'}],
            'groups': [{'term': '1', 'pattern': '----', 'covers': list(range(CELL_COUNT))}]
        }

    current = {to_bin(t): {t} for t in all_terms}
    prime_implicants = {}

    while current:
        next_level = {}
        used = set()
        items = list(current.items())

        for i in range(len(items)):
            p1, m1 = items[i]
            for j in range(i + 1, len(items)):
                p2, m2 = items[j]
                combined = combine_terms(p1, p2)
                if combined:
                    merged = m1 | m2
                    if combined in next_level:
                        next_level[combined] |= merged
                    else:
                        next_level[combined] = merged
                    used.add(p1)
                    used.add(p2)

        for p, m in current.items():
            if p not in used and (m & actual):
                prime_implicants[p] = m

        current = next_level

    if not prime_implicants:
        return {
            'expression': '0',
            'kmap': build_kmap(minterms, dont_cares),
            'steps': [],
            'groups': []
        }

    uncovered = set(actual)
    selected = set()

    for mint in list(uncovered):
        covering = [p for p, m in prime_implicants.items() if mint in m]
        if len(covering) == 1:
            p = covering[0]
            selected.add(p)
            uncovered -= prime_implicants[p]

    while uncovered:
        best_p = None
        best_count = 0
        for p, m in prime_implicants.items():
            if p in selected:
                continue
            count = len(m & uncovered)
            if count > best_count:
                best_count = count
                best_p = p
        if best_p is None:
            break
        selected.add(best_p)
        uncovered -= prime_implicants[best_p]

    terms = []
    for p in sorted(selected, key=lambda x: (x.count('-'), x)):
        terms.append(get_term_name(p))
    expression = ' + '.join(terms)

    groups = []
    for p in sorted(selected, key=lambda x: (x.count('-'), x)):
        groups.append({
            'term': get_term_name(p),
            'pattern': p,
            'covers': sorted(prime_implicants[p])
        })

    steps = []

    minterm_str = ', '.join(f'm{n}({to_bin(n)})' for n in sorted(minterms))
    steps.append({'title': 'Minterms (1s)', 'text': f'Cells set to 1: {minterm_str}'})

    if dont_cares:
        dc_str = ', '.join(f'm{n}' for n in sorted(dont_cares))
        steps.append({'title': "Don't Cares (Xs)", 'text': f'Cells with don\'t care: {dc_str}'})

    pi_list = []
    for p in sorted(prime_implicants, key=lambda x: (x.count('-'), x)):
        pi_list.append({
            'pattern': p,
            'term': get_term_name(p),
            'covers': sorted(prime_implicants[p])
        })
    steps.append({'title': 'Prime Implicants', 'text': f'Found {len(prime_implicants)} prime implicant(s)', 'implicants': pi_list})

    sel_list = []
    for p in sorted(selected, key=lambda x: (x.count('-'), x)):
        sel_list.append({
            'pattern': p,
            'term': get_term_name(p),
            'covers': sorted(prime_implicants[p])
        })
    steps.append({'title': 'Minimal Cover', 'text': f'Selected {len(selected)} term(s) for minimal expression', 'selected': sel_list})

    return {
        'expression': expression,
        'kmap': build_kmap(minterms, dont_cares),
        'steps': steps,
        'groups': groups
    }

File: static/js/solver.js
const KmapSolver = (function () {

  function getConfig(kmapType) {
    if (kmapType === '3var-ab-c') {
      return {
        variables: ['A', 'B', 'C'],
        nVars: 3,
        cellCount: 8,
        rows: ['00', '01', '11', '10'],
        cols: ['0', '1'],
        cornerLabel: 'AB\\C'
      };
    } else if (kmapType === '3var-a-bc') {
      return {
        variables: ['A', 'B', 'C'],
        nVars: 3,
        cellCount: 8,
        rows: ['0', '1'],
        cols: ['00', '01', '11', '10'],
        cornerLabel: 'A\\BC'
      };
    } else {
      return {
        variables: ['A', 'B', 'C', 'D'],
        nVars: 4,
        cellCount: 16,
        rows: ['00', '01', '11', '10'],
        cols: ['00', '01', '11', '10'],
        cornerLabel: 'AB\\CD'
      };
    }
  }

  function toBin(n, nVars) {
    return n.toString(2).padStart(nVars, '0');
  }

  function getTermName(pattern, variables) {
    const parts = [];
    for (let i = 0; i < pattern.length; i++) {
      const ch = pattern[i];
      if (ch === '1') {
        parts.push(variables[i]);
      } else if (ch === '0') {
        parts.push(variables[i] + "'");
      }
    }
    return parts.length ? parts.join('') : '1';
  }

  function combineTerms(t1, t2, nVars) {
    let diffIdx = null;
    const result = [];
    for (let i = 0; i < nVars; i++) {
      const a = t1[i];
      const b = t2[i];
      if (a === '-' && b === '-') {
        result.push('-');
      } else if (a === '-' || b === '-') {
        return null;
      } else if (a === b) {
        result.push(a);
      } else {
        if (diffIdx !== null) return null;
        diffIdx = i;
        result.push('-');
      }
    }
    return diffIdx !== null ? result.join('') : null;
  }

  function setUnion(a, b) {
    return new Set([...a, ...b]);
  }

  function setIntersection(a, b) {
    return new Set([...a].filter(x => b.has(x)));
  }

  function setDifference(a, b) {
    return new Set([...a].filter(x => !b.has(x)));
  }

  function buildKmap(minterms, dontCares, kmapType) {
    const config = getConfig(kmapType);
    const kmap = [];
    for (const r of config.rows) {
      for (const c of config.cols) {
        const bits = r + c;
        const idx = parseInt(bits, 2);
        let val;
        if (dontCares.has(idx)) {
          val = 'x';
        } else {
          val = minterms.has(idx) ? 1 : 0;
        }
        kmap.push({ index: idx, binary: bits, value: val, row: r, col: c });
      }
    }
    return kmap;
  }

  function simplify(mintermsArr, dontCaresArr, kmapType) {
    const minterms = new Set(mintermsArr);
    const dontCares = new Set(dontCaresArr);
    const config = getConfig(kmapType);
    const nVars = config.nVars;
    const cellCount = config.cellCount;
    const variables = config.variables;

    const allTerms = [...new Set([...minterms, ...dontCares])].sort((a, b) => a - b);
    const actual = new Set(minterms);

    if (actual.size === 0) {
      return {
        expression: '0',
        kmap: buildKmap(minterms, dontCares, kmapType),
        steps: [{ title: 'No Minterms', text: 'No minterms selected — output is 0' }],
        groups: []
      };
    }

    if (actual.size === cellCount) {
      const pattern = '-'.repeat(nVars);
      return {
        expression: '1',
        kmap: buildKmap(minterms, dontCares, kmapType),
        steps: [{ title: 'All Minterms', text: 'All cells are 1 — output is 1' }],
        groups: [{ term: '1', pattern, covers: [...Array(cellCount).keys()] }]
      };
    }

    let current = new Map();
    for (const t of allTerms) {
      current.set(toBin(t, nVars), new Set([t]));
    }

    const primeImplicants = new Map();

    while (current.size > 0) {
      const nextLevel = new Map();
      const used = new Set();
      const items = [...current.entries()];

      for (let i = 0; i < items.length; i++) {
        const [p1, m1] = items[i];
        for (let j = i + 1; j < items.length; j++) {
          const [p2, m2] = items[j];
          const combined = combineTerms(p1, p2, nVars);
          if (combined !== null) {
            const merged = setUnion(m1, m2);
            if (nextLevel.has(combined)) {
              nextLevel.set(combined, setUnion(nextLevel.get(combined), merged));
            } else {
              nextLevel.set(combined, merged);
            }
            used.add(p1);
            used.add(p2);
          }
        }
      }

      for (const [p, m] of current) {
        if (!used.has(p) && setIntersection(m, actual).size > 0) {
          primeImplicants.set(p, m);
        }
      }

      current = nextLevel;
    }

    if (primeImplicants.size === 0) {
      return {
        expression: '0',
        kmap: buildKmap(minterms, dontCares, kmapType),
        steps: [],
        groups: []
      };
    }

    let uncovered = new Set(actual);
    const selected = new Set();

    for (const mint of [...uncovered]) {
      const covering = [];
      for (const [p, m] of primeImplicants) {
        if (m.has(mint)) covering.push(p);
      }
      if (covering.length === 1) {
        const p = covering[0];
        selected.add(p);
        uncovered = setDifference(uncovered, primeImplicants.get(p));
      }
    }

    while (uncovered.size > 0) {
      let bestP = null;
      let bestCount = 0;
      for (const [p, m] of primeImplicants) {
        if (selected.has(p)) continue;
        const count = setIntersection(m, uncovered).size;
        if (count > bestCount) {
          bestCount = count;
          bestP = p;
        }
      }
      if (bestP === null) break;
      selected.add(bestP);
      uncovered = setDifference(uncovered, primeImplicants.get(bestP));
    }

    const patternSort = (a, b) => {
      const aD = (a.match(/-/g) || []).length;
      const bD = (b.match(/-/g) || []).length;
      return aD !== bD ? aD - bD : a.localeCompare(b);
    };

    const sortedSelected = [...selected].sort(patternSort);

    const terms = sortedSelected.map(p => getTermName(p, variables));
    const expression = terms.join(' + ');

    const groups = sortedSelected.map(p => ({
      term: getTermName(p, variables),
      pattern: p,
      covers: [...primeImplicants.get(p)].sort((a, b) => a - b)
    }));

    const steps = [];

    const mintermStrs = [...minterms].sort((a, b) => a - b).map(n => `m${n}(${toBin(n, nVars)})`);
    steps.push({ title: 'Minterms (1s)', text: `Cells set to 1: ${mintermStrs.join(', ')}` });

    if (dontCares.size > 0) {
      const dcStrs = [...dontCares].sort((a, b) => a - b).map(n => `m${n}`);
      steps.push({ title: "Don't Cares (Xs)", text: `Cells with don't care: ${dcStrs.join(', ')}` });
    }

    const sortedPis = [...primeImplicants.keys()].sort(patternSort);
    const piList = sortedPis.map(p => ({
      pattern: p,
      term: getTermName(p, variables),
      covers: [...primeImplicants.get(p)].sort((a, b) => a - b)
    }));
    steps.push({ title: 'Prime Implicants', text: `Found ${primeImplicants.size} prime implicant(s)`, implicants: piList });

    const selList = sortedSelected.map(p => ({
      pattern: p,
      term: getTermName(p, variables),
      covers: [...primeImplicants.get(p)].sort((a, b) => a - b)
    }));
    steps.push({ title: 'Minimal Cover', text: `Selected ${selected.size} term(s) for minimal expression`, selected: selList });

    return {
      expression,
      kmap: buildKmap(minterms, dontCares, kmapType),
      steps,
      groups
    };
  }

  return { simplify };
})();